### PR TITLE
feat(including bugs): minor changes that increases functionality

### PR DESCRIPTION
**Pull request by im-coder-lg**
Hi @sumeshir26,

This is a really great project! Although it has a few glitches(listed below), it's a chance for me to learn some new things too. So, after long research(and switching between computers), I have made a series of minor changes in the source file(main.py) and removed the spec listing from the .gitignore(later on that). I have also generated a newer executable with the TimerX icon you provided in the `assets` directory. And, I think you don't need the `setup.py` file anymore since PyInstaller can build everything for you(more on that later).

Now, the glitches are as follows:

1. You need to click the Play button twice to start the timer.
2. The timer takes one second to start and the rest 5 seconds as countdown, so reducing the extra second seems like a good idea.
3. This has been happening for me, but whenever the timer ends, I get a driver error(it doesn't play the audio but my driver's all cool, I can see YT videos, and hear audio content). Does the same happen for you too?

Next, why did I remove the spec file from the .gitignore?

I think you used Pyinstaller the first time you built it. And there is an additional tool with PyInstaller called `pyi-makespec`, which generates SPEC files for making building the app easier. Basically, clone the repo, get PyInstaller, and run this: `pyinstaller main.spec` in the root. You get an absolutely smashing executable with the TimerX logo(put it in the root dir to run that, so that it will detect the assets), and it has no consoles open(although, a console might open during errors like running the app from the `dist` dir due to assets).

So, I'm done here.  I need your opinions on this, and also about the driver error I mentioned above.

### CODE REVIEW

**Comment by sumeshir26** [comment]
@im-coder-lg no, I do not get any driver errors🤔🤔

**Comment by sumeshir26** [comment]
❤️Awsome work! Do you want me to merge?

**Comment by im-coder-lg** [comment]
> ❤️Awsome work! Do you want me to merge?

Well not yet because I am still analyzing the code... The icon seems zoomed in when it's a plain exe... Need to check that, and can you try sharing the original TimerX icon file? Maybe I can try making it into an .ico file and test with that...

> @im-coder-lg no, I do not get any driver errors🤔🤔

As expected because you run Windows 11 which has advanced drivers, so I think this could be the time to upgrade to Linux soon... But other than that, great! I just need to find how to make macOS builds and add that here so that Mac users don't need to feel loft alone or also they won't need to install Wine(compatibility layer) to run TimerX...  Will see that tomorrow.

**Comment by im-coder-lg** [comment]
Do you know anyone who has a Mac @sumeshir26? Maybe you can test TimerX's Mac build on that. It seems that you can use py2exe as well as PyInstaller, so if we could get a hand at macOS, we could try building TimerX from the terminal. If it succeeds, we are good on a Mac app file.

**Comment by im-coder-lg** [comment]
The Linux build is failing. I have a Raspberry Pi running Raspbian Buster(Debian 10 for Raspberry Pi) and tried building it. The build started well, but the file doesn't execute. I mean, in Geany, if you execute main.py, it'll throw up an error saying that `assets/logo.ico` isn't found. Linux doesn't use .ico files, so we need to find a proper icon file for Linux. Think macOS has this same error?

**Comment by im-coder-lg** [comment]
[asciicast](https://asciinema.org/a/448579)

The video is called an Asciicast, used for showing console logs as videos. On the Pi, I have configured it to show a Neofetch result at the beginning to show the OS. That's proof it's Raspbian. After that, I changed the directory to TimerX's dir, showed my branch(fork not the base TimerX), and entered `python3 main.py` and I ended up with the error. Can you try sharing the original logo image? Maybe I can make macOS and Linux icons from that.

**Comment by sumeshir26** [comment]
@im-coder-lg the original logo is in the assets/images dir

**Comment by im-coder-lg** [comment]
We got some issues.

1. The .ico code works only on Windows, which means we need different files+icons for each OS.
2. The logo doesn't work as a JPEG or as a JPG, it works as a PNG, and the PNG I generated was low-quality. Making an icon is the best idea.
3. Your about window's icon doesn't work due to the JPEG error, so we need different icons and different files for each OS, which makes 3 main.py files, 3 configurator.py files and 3 main.spec files, which squares the file load.

We need a solution to this soon. I could change the ABout page's icon to a bitmap icon(.ico) but Macs use .icns and Linux uses .icon(I guess). How can we fix this?

**Comment by im-coder-lg** [comment]
I just hopped onto the GAUDC project(that gave me a lot of ideas to implement here) and saw the [app/res/icons](https://github.com/Melvin-Abraham/Google-Assistant-Unofficial-Desktop-Client/tree/master/app/res/icons) folder and I saw a _horde_ of icons! And as expected, there was an ICNS. So, Linux seems to have PNGs or SVG images, but you need an ICNS for Mac. So, we hit something here, and maybe I could get a fresh ICNS for Mac. @sumeshir26 does Python recognize OSes? Like, if I run TimerX from source on Linux, will the file  know that it runs on Linux? If it does, we could use `if` and `elif` commands to manage the icon issue. And also, if the system is neither Windows nor MacOS or Linux, it will resort to the PNG(for now). Is this possible?

**Comment by im-coder-lg** [comment]
We could use [this](https://stackoverflow.com/questions/8220108/how-do-i-check-the-operating-system-in-python) solution I found in Stack Overflow:
```python
from sys import platform
if platform == "linux" or platform == "linux2":
    # linux
elif platform == "darwin":
    # OS X
elif platform == "win32":
    # Windows...
```
We also need a `requirements.txt` so that users get the deps like so: `pip install -r rquirements.txt`

**Comment by sumeshir26** [comment]
@im-coder-lg Let me add it real quick..

**Comment by im-coder-lg** [comment]
Wait what? You are adding the code? We need the icon files! An ICNS, a Linux icon file and an improved PNG(no more JPEG, it throws up an error)

**Comment by im-coder-lg** [comment]
It is hard to make an ICNS file. I have used online converters and always fail under that.

**Comment by sumeshir26** [comment]
@im-coder-lg Let me try to make a icns

**Comment by im-coder-lg** [comment]
Does py2exe support SPEC files? If they do, we could use py2xe for the generation of files and pyinstaller for modifying the SPEC file. Could that be possible?

**Comment by im-coder-lg** [comment]
Mac failed. Maybe we need to use a different build platform for Macs. What about PyInstaller?

**Comment by im-coder-lg** [comment]
@sumeshir26 jackpot. I made the ICNS using an online tool, it's on my device, uploading here now.

**Comment by im-coder-lg** [comment]
The thing is since it was made from the PNG, how'd the quality be?

**Comment by sumeshir26** [comment]
@im-coder-lg The build fails on line 63 of setup.py

**Comment by sumeshir26** [comment]
Anyway, i got the icns

**Comment by sumeshir26** [comment]
Uploding in the assets directory on your fork...

**Comment by im-coder-lg** [comment]
I was about to do the same thing! Hope you made it from the JPEG, because the PNG looked strange as an icon(during my tests)  but anyway, we got the ICNS, we need a Linux icon file.

**Comment by sumeshir26** [comment]
@im-coder-lg I dont hawe write access

**Comment by im-coder-lg** [comment]
image
I get this

Maybe you force-push(`git push origin branch --force`). Change `origin` and `branch` to my repo origin(`git remote add im-coder-lg https://github.com/im-coder-lg/TimerX.git`) and checkout to `upgrade`

**Comment by sumeshir26** [comment]
@im-coder-lg As far as I know, linux uses png

**Comment by im-coder-lg** [comment]
> @im-coder-lg As far as I know, linux uses png

Same. We need a higher photo quality of the logo.

**Comment by im-coder-lg** [comment]
I sent a collaborator invite to you on my fork, maybe you can accept, switch branches and push the ICNS.

**Comment by sumeshir26** [comment]
image

**Comment by im-coder-lg** [comment]
maybe MS Paint had a part of messing this up. I need another PNG and some time. Will report here after Raspbian test.

**Comment by sumeshir26** [comment]
Im sorry, pushed to the wrong branch

**Comment by im-coder-lg** [comment]
Woah, Linux succeeds, Windows stops and MacOS fails. Kinda similar to my checks upgrade(see Actions tab in my fork)

**Comment by sumeshir26** [comment]
image

**Comment by sumeshir26** [comment]
@im-coder-lg You could possibly test it through WSL

**Comment by im-coder-lg** [comment]
Problem is, my Windows10 PC has a problem, and my only shot, Gitpod is kinda failing.

**Comment by sumeshir26** [comment]
Let me try

**Comment by sumeshir26** [comment]
@im-coder-lg Buut... dont we need a .deb or .rpm for linux?

**Comment by im-coder-lg** [comment]
Can't we make it? We could use a third-party executable creator for Linux(Mac+Win needed, can be easily created using the same tool if it allows) and make .deb and .rpm files, right?

**Comment by im-coder-lg** [comment]
Or we could create an AppImage file, they seem to work on Linux.

**Comment by sumeshir26** [comment]
I found the problem... The build and the dist folders where in the gitignore!

**Comment by im-coder-lg** [comment]
> @im-coder-lg The build fails on line 63 of setup.py

About this, well...
https://github.com/im-coder-lg/TimerX/blob/68bdaf5038b6633d271b74496c7ecc6fe686dea0/setup.py#L1-L4
The code is literally four lines, how would it be 63+ lines? Unless the issue is in `main.py`...

**Comment by im-coder-lg** [comment]
> I found the problem... The build and the dist folders where in the gitignore!

Also about this, how would it affect TimerX? We just need to confirm if the build works. I know that Windows works, Linux build does work but the file doesn't(yet) and more issues.

**Comment by im-coder-lg** [comment]
Mac, idk, I need to push changes, made one on my Raspberry Pi(platform icon config) but shut it down before pushing it here.

**Comment by im-coder-lg** [comment]
Force-pushed here so that Raspbian ignores issues and just pushes the change.

File: main.py [hunk 1]
```diff
@@ -5,13 +5,22 @@
...
 import configurator
+from sys import platform
 
 # TKINTER WINDOW
 app = Tk()
 app.title('TimerX')
 app.geometry('300x200')
 app.iconbitmap(r"./assets/logo.ico")
 
+if platform == "linux" or platform == "linux2":
```
Comment (im-coder-lg) [inline main.py:16]: doesn't seem to work as expected

Comment (sumeshir26) [reply main.py:16]: Remove line 14

**Comment by im-coder-lg** [comment]
The main.py platform code works, but configurtor.py doesn't. Need to investigate.

**Comment by sumeshir26** [comment]
image
The taskbar icon does not change😕

**Comment by im-coder-lg** [comment]
Same, but I think I need to try building TimerX from the upgrade branch.

**Comment by im-coder-lg** [comment]
It works in the build. Try building TimerX using PyInstaller.

**Comment by sumeshir26** [comment]
The macos build fails?

**Review by sumeshir26 — APPROVED** [review]
Looks Good!

**Comment by im-coder-lg** [comment]
> The macOS build fails?

Guess so. If Mac users don't get TimerX, they could use compatibility layers. Hoped to get a Mac build but don't have a Mac... Does Mac use AppImage? If it does, we could use one Linux build for the Mac, that'd work. But the next hurdle is the M1 processors, but it uses ARM. So, if Macs support AppImage, we need an AMD(Intel?) and an ARM(Apple Silicon) build. @sumeshir26 about the About page, why can't we use a separate file? The file doesn't recognise the PhotoImage module.

**Comment by im-coder-lg** [comment]
We need to fix one last thing for now. Configurator.py split. We need the about window to be separate or it should recognise what a PhotoImage line is.

**Comment by sumeshir26** [comment]
@im-coder-lg I'll try after I get the dark mode working

**Comment by im-coder-lg** [comment]
The icon can be used from the Material Icons library(open-source here on GitHub).

**Comment by sumeshir26** [comment]
@im-coder-lg What icon?

**Comment by sumeshir26** [comment]
I think I will just use darkdetct to detect the system theme

**Comment by im-coder-lg** [comment]
> @im-coder-lg What icon?

Dark mode toggle. If you use an icon instead of text, it'd just blend into the layout(not meaning "hide", it'd be like it _belongs_ there). Make sure to use a lighter shade of black so that users can see things properly.

**Comment by sumeshir26** [comment]
@im-coder-lg What OS are you on?

**Comment by im-coder-lg** [comment]
Windows 7, but I am hoping to switch to Zorin OS within the next few weeks. I also have a Raspberry Pi running Raspbian Bullseye, the Linux tests I said were from that.

**Comment by im-coder-lg** [comment]
Windows 7, but I am hoping to switch to Zorin OS within the next few weeks. I also have a Raspberry Pi running Raspbian Bullseye, the Linux tests I said were from that.

**Comment by sumeshir26** [comment]
@im-coder-lg Should I merge?

**Comment by sumeshir26** [comment]
Idk anyone with a mac

**Comment by im-coder-lg** [comment]
> @im-coder-lg Should I merge?

Not yet

> Idk anyone with a mac

Same. I hate getting a Hackintosh as a VM, so we need to make Windows and Linux builds only, till you find a person using a Mac. I know I can't find one, but I will try.

**Comment by sumeshir26** [comment]
What else do you want to add?

**Comment by sumeshir26** [comment]
@im-coder-lg

**Comment by im-coder-lg** [comment]
Sorry man, I was too busy+had headache, I will push the changes by tonight.

**Comment by sumeshir26** [comment]
@im-coder-lg No problem 😉 Get well soon…

**Comment by im-coder-lg** [comment]
> @im-coder-lg No problem 😉 Get well soon…

Thanks man, got rid of them, happens, now I am ready to get the PR for merge 😀

**Comment by im-coder-lg** [comment]
Whoa. Never expected to see the fact that, once I made breaking changes to configuration.py, and thought I pushed the changes here, the PR is ready for erge

**Comment by im-coder-lg** [comment]
Was

**Comment by sumeshir26** [comment]
I don’t see any changes?

**Comment by im-coder-lg** [comment]
Well because there are no changes. I thought I made some breaking changes so I asked for time.

The PR is ready for merge.

**Review by sumeshir26 — APPROVED** [review]

**Pull request merged**